src/metabase/upload.clj

```diff
@@ -311,17 +338,21 @@
   "Loads a table from a CSV file. If the table already exists, it will throw an error.
    Returns the file size, number of rows, and number of columns."
   [driver db-id table-name ^File csv-file]
-  (let [col->upload-type   (detect-schema csv-file)
-        col->database-type (update-vals col->upload-type (partial driver/upload-type->database-type driver))
-        column-names       (keys col->upload-type)]
-    (driver/create-table! driver db-id table-name col->database-type)
+  (let [{col-to-insert->upload-type :extant-columns
+         gen-col->upload-type       :generated-columns} (detect-schema csv-file)
+        col-to-create->col-spec                         (upload-type->col-specs driver
```

Comment:
I'm wondering if we should let `detect-schema` return a single array of `[[col-name, col-type]]` like before.

One problem is  `extant-columns` and `generated-columns` can contain duplicated keys when a csv has an `id` column with `varchar` type.

in that case, `extant-columns` will have an `id` column with varchar type and `generated-columns` has an `id` column with auto-incremented-id type.

This makes the logic separated I think,  because `detect-schema` is no longer returns the "final" schema, it's up to `load-from-csv!` to produce the final schema by doing `(merge gen-col->upload-type col-to-insert->upload-type)` <<- notice this merge is order-important as well.

That's said what's the benefit of having `detect-schema` to return a map like that?

Reply:
Example

with this schema

```
(csv-file-with ["id,ship,captain"
                "a,Serenity,Malcolm Reynolds"
                 "b,Millennium Falcon,Han Solo"])
```
 `col-to-create->col-spec` is `{:captain [[:varchar 255]], :id [[:varchar 255]], :ship [[:varchar 255]]}`

and returned from `load-from-csv!`

```
{:generated-columns 1, :num-columns 3, :num-rows 2, :size-mb 6.866455078125E-5}
```

this report is wrong because there were 0 columns that were auto-generated.

Reply:
Good catch. The data structure containing the column names + type info returned by `detect-schema` imo should be a single ordered-map to make it clearer that the column names are still unique. But if you do that, the return value will still need to include information about which columns will be automatically generated. Maybe it could return a shape with the keys 
`:col-name->type :: OrderedMap, :autogenerated-col-names :: Set` and then you could derive the col names to insert with `(apply dissoc col-name->type autogenerated-col-names)`

Reply:
Here's the string PR: #34592

Reply:
On b), detecting string columns named ID will not solve the problem _entirely_, because there can be columns called ID or PK etc that are other types, like dates. It works but I don't like that it depends on the merge being in the right order with `(merge gen-col->upload-type col-to-insert->upload-type)`.

Anyway its not a strong reason to block this PR so lets ship it